src/coprocessor/codec/row_codec/v_128.rs

```diff
@@ -0,0 +1,584 @@
+            Ok(())
+        }
+        Datum::Json(j) => {
+            dest.encode_json(&j).expect("TODO: handle this");
+            Ok(())
+        }
+        Datum::F64(_) | Datum::Time(_) | Datum::Dur(_) | Datum::Null => unreachable!(),
+        _ => unimplemented!(),
+    }
+}
+
+#[allow(clippy::match_overlapping_arm)]
+fn encode_i64(v: i64, dest: &mut Vec<u8>) -> Result<()> {
+    match v {
+        -128..=127 => {
+            dest.push(v as i8 as u8);
+            Ok(())
+        }
+        -32768..=32767 => dest.write_i16_le(v as i16),
+        -2_147_483_648..=2_147_483_647 => dest.write_i32_le(v as i32),
+        _ => dest.write_i64_le(v),
+    }
+}
+
+fn encode_u64(v: u64, dest: &mut Vec<u8>) -> Result<()> {
+    match v {
+        0..=255 => {
+            dest.push(v as u8);
+            Ok(())
+        }
+        256..=65535 => dest.write_u16_le(v as u16),
+        65536..=4_294_967_295 => dest.write_u32_le(v as u32),
+        _ => dest.write_u64_le(v),
+    }
+}
+
+fn encode_bytes(b: Vec<u8>, dest: &mut Vec<u8>) -> Result<()> {
+    dest.extend_from_slice(&b);
+    Ok(())
+}
+
+//    Float = 4,
+//    Null = 6,
+//    NewDate = 14,
+//    Bit = 16,
+//    Enum = 0xf7,
+//    Set = 0xf8,
+//    Geometry = 0xff,
+fn decode_datum(col_info: &ColumnInfo, ctx: &EvalContext, data: &[u8]) -> Result<Datum> {
+    use crate::coprocessor::codec::mysql::{Decimal, Duration as MysqlDuration, Json, Time};
+    use cop_datatype::{FieldTypeFlag, FieldTypeTp::*};
+    use core::convert::TryInto;
+    match col_info.tp() {
+        Tiny | Short | Long | LongLong | Int24 | Year => {
+            if col_info.flag().contains(FieldTypeFlag::UNSIGNED) {
+                Ok(Datum::U64(decode_u64(data)))
+            } else {
+                Ok(Datum::I64(decode_i64(data)))
+            }
+        }
+        VarChar | VarString | String | TinyBlob | MediumBlob | LongBlob | Blob => {
+            Ok(Datum::Bytes(data.to_vec()))
+        }
+        Double | Float => Ok(Datum::F64(f64::from_bits(decode_u64(data)))),
+        Date | DateTime | Timestamp => {
+            let t = Time::from_packed_u64(
+                decode_u64(data),
+                col_info.tp().try_into().unwrap(),
+                col_info.decimal() as i8,
+                &ctx.cfg.tz,
+            )
+            .expect("TODO: handle this");
+            Ok(Datum::Time(t))
+        }
+        NewDecimal => {
+            let d = Decimal::decode(&mut data.to_vec().as_slice()).expect("TODO: handle this");
+            Ok(Datum::Dec(d))
+        }
+        JSON => {
+            let j = Json::decode(&mut data.to_vec().as_slice()).expect("TODO: handle this");
+            Ok(Datum::Json(j))
+        }
+        Duration => {
+            let d = MysqlDuration::from_nanos(decode_i64(data), col_info.decimal() as i8)
+                .expect("TODO: handle this");
+            Ok(Datum::Dur(d))
+        }
+        _ => unimplemented!(),
+    }
+}
+
+fn decode_u64(data: &[u8]) -> u64 {
```

Comment:
Prefer to provide interface over BufferReader and BufferWriter to make it more friendly to use.

Actually the `codec` crate is implemented as 2 layers:
layer 1: the main implementation, over fixed size buffers (`&[u8]` or `&mut [u8]`).
layer 2: the high level encapsulation, based on layer 1 and BufferReader/Writer interface, to make it more friendly.